### PR TITLE
ENSIP-19 Updates (2)

### PR DESCRIPTION
**Pull request by shrugs**
supercedes https://github.com/namehash/ensnode/pull/871 because of merge conflicts

closes https://github.com/namehash/ensnode/issues/868

- [x] add `acceleration` config param to sdk/hooks
- [x] move acceleration logic files into `lib/acceleration` for organization
- [x] integrate new default public resolver with address defaulting behavior `0xF29100983E058B709F3D539b0c765937B804AC15` (mainnet) and make sure it's indexed as part of reverse-resolvers plugin, right?
- [x] make sure that UniversalResolver's `offset` parameter is being interpreted correctly — should return 
- [x] batch reverse resolution api endpoint + sdk + hooks support
- [x] define entity representing `(chainId, registrarAddress, address) -> Name`
- [x] upsert those entities in `apps/ensindexer/src/handlers/multi-chain/StandaloneReverseRegistrar.ts`
- [x] integrate a new acceleration pathway
  - option 1: if the `activeResolver` is one of the well-known `ChainReverseResolver` addresses then we know the `chainId` and the `registrarAddress` to use: just go ahead and fetch the `name` record via index via those entities and return it, making sure to default as necessary
  - option 2: inspect the name being resolved and if it is of the acceleratable on-chain static registrar-backed reverse resolvers (`default.reverse`, `[coinType].reverse`) then accelerate it by looking up in the index
  - depends on whether we want to hard-code the resolvers and check against the result of findResolver (ala current acceleration) or whether we want to accelerate based on the name itself. name itself is simpler but going by resolver is more exact. but fragile if the resolvers would ever change
- [ ] ~~rename the plugin to `reverse-resolution` instead of `reverse-resolvers` (inc all datasource names)~~
- [x] disallow chainId 0 from being selected via the api
- [x] rename all the batch stuff to multichain 
- [x] update primary-name(s) examples with an account that has different primary names per address (ideally with default) and show full expected response format
- [x] understand normalization behavior (whether emitted names are guaranteed to be normalized)
- [x] understand encoded label hash handling
- [x] understand how NameCoder `offset` works wrt unicode points
- [x] `http://localhost:42069/api/resolve/records/jesse.base.eth?addresses=2147492101` is returning `0x` but that should be `null`
- [x] changesets

### CODE REVIEW

File: packages/ensnode-sdk/src/ens/constants.ts [hunk 1]
```diff
@@ -10,17 +11,3 @@ export const ROOT_NODE: Node = namehash("");
  * See apps/ensindexer/src/handlers/Registry.ts for context.
  */
 export const REVERSE_ROOT_NODES: Set<Node> = new Set([namehash("addr.reverse")]);
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/ensnode-sdk/src/ens/constants.ts:13]: https://github.com/namehash/ensnode/issues/903

File: packages/datasources/src/mainnet.ts [hunk 2]
```diff
@@ -259,101 +263,119 @@ export default {
...
    */
   [DatasourceNames.ReverseResolverBase]: {
     chain: base,
     contracts: {
-      ReverseResolver: {
-        abi: ResolverConfig.abi,
-        // NOTE: this is basenames L2Resolver, NOT ENSIP-19 Compliant
-        address: "0xC6d566A56A1aFf6508b41f6c90ff131615583BCD", // TODO: update this address
-        startBlock: 17575714,
+      // TODO: i think we can simply delete this after basenames ENSIP-19 migration
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/datasources/src/mainnet.ts:306]: done, they will do a full migration, so legacy base contracts can be ignored and removed. will confirm with ens team for linea

File: packages/datasources/src/namespaces.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ */
+export const getENSRootChain = (namespaceId: ENSNamespaceId) =>
+  getDatasource(namespaceId, DatasourceNames.ENSRoot).chain;
+
+/**
+ * Returns the chain id for the ENS Root Datasource within the selected namespace.
+ *
+ * @returns the chain ID that hosts the ENS Root
+ */
+export const getENSRootChainId = (namespaceId: ENSNamespaceId) => getENSRootChain(namespaceId).id;
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/datasources/src/namespaces.ts:56]: in this case we're specifically using the inferred types

File: apps/ensindexer/src/api/lib/batch-reverse-resolution.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @param address the adddress whose Primary Names to resolve
+ * @param chainIds the set of chainIds within which to resolve the address' Primary Name
+ */
+export async function batchResolveReverse(
+  address: Address,
+  chainIds: ChainId[] | undefined,
+  options: { accelerate?: boolean } = { accelerate: true },
+): Promise<PrimaryNames> {
+  const _chainIds = chainIds || CHAIN_IDS_BY_NAMESPACE[config.namespace];
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensindexer/src/api/lib/batch-reverse-resolution.ts:40]: if outside the set, reverse resolution for that chain id would proceed as normal — My understanding is that it's not illegal in the protocol to ask for a chain ID that isn't explicitly supported by a specific coin type resolver. It'll just fall back to the default resolver, which is intended. The well-known chain IDs just happen to have a specific behavior, which is that they post confirmations to L1 and can be verified on-chain and returned synchronously by using the special reverse resolvers for their respective chain ids

File: apps/ensindexer/src/api/lib/find-resolver.ts [hunk 5]
```diff
@@ -24,67 +31,74 @@ const NULL_RESULT: FindResolverResult = {
...
-  // .linea.eth -> ensroot.LineanamesL1Resolver.address;
-  // note that with that acceleration approach we may need to explicitly not suppport or make a
-  // carve-out for those base.eth subdomains on mainnet
-
-  // Implicit Invariant: findResolver is _always_ called for the ENSRoot Chain and then _ONLY_
-  // called with chains for which we are guaranteed to have the Domain-Resolver relations indexed.
-  // This is enforced by the requirement that `forwardResolve` with non-ENSRoot chain ids is only
-  // called when an known offchain lookup resolver defers to a plugin that is active.
+/**
+ * Identifies the active resolver on `chainId` for `name`.
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensindexer/src/api/lib/find-resolver.ts:37]: we specifically use 'active' resolver throughout the codebase for this concept. iirc we discussed active vs effective months ago. currently the word effective is used in the context of 'effective owner' of a domain.

File: apps/ensindexer/src/api/lib/forward-resolution.ts [hunk 6]
```diff
@@ -118,25 +122,24 @@ async function _resolveForward<SELECTION extends ResolverRecordsSelection>(
...
           // but honestly the state drift is at max 1 block on L1 and a block or two on an L2, it's pretty negligible,
           // so maybe we just ignore this issue entirely
-          const normalizedName = withSpan(tracer, "normalization", {}, () => normalize(name));
+          const normalizedName = normalize(name);
           if (name !== normalizedName) {
             throw new Error(`Name "${name}" must be normalized ("${normalizedName}").`);
           }
 
           // TODO: more name normalization logic (return values of `name` record for example)
           // TODO: need to handle encoded label hashes in name, yeah?
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensindexer/src/api/lib/forward-resolution.ts:131]: deferring to https://github.com/namehash/ensnode/issues/829

File: apps/ensindexer/src/api/lib/reverse-resolution.ts [hunk 7]
```diff
@@ -42,130 +41,120 @@ export async function resolveReverse(
...
-          if (!records.name) {
+          if (!name) {
+            // If name is empty, no primary name exists for this address.
             span.setStatus({
               code: SpanStatusCode.ERROR,
-              message: `No Primary Name for coinType "${evmChainIdToCoinType(chainId)}" or default coinType.`,
+              message: `No Primary Name for coinType "${coinType}" or default coinType.`,
             });
             return null;
           }
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensindexer/src/api/lib/reverse-resolution.ts:76]: deferred to https://github.com/namehash/ensnode/issues/829

File: packages/ensnode-schema/src/primary-names.schema.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import { onchainTable, relations, uniqueIndex } from "ponder";
+import { account } from "./subgraph.schema";
+
+// add the additional relationships to subgraph's Account entity
+export const ext_primaryNames_domain_relations = relations(account, ({ one, many }) => ({
+  // account has many primary names
+  primaryNames: many(ext_primaryName),
+}));
+
+// tracks an Account's Primary Name by CoinType
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/ensnode-schema/src/primary-names.schema.ts:14]: for all ensip-19 supported chains, yes, because they're guaranteed to be onchain and are well-known. this does not (and cannot) materialize the primary name from `[addr].addr.reverse` resolvers because those can be any contract. those lookup pathways are accelerated as normal though (if using the default public resolver), so in practice the cost of looking up all ensip-19 primary names for an address is very fast (60ms) because each `resolveForward` is fully accelerated

[the remark this answers is not in the document]
Comment (shrugs) [reply packages/ensnode-schema/src/primary-names.schema.ts:14]: i'll add a comment that this is not a fully materialization and cannot be used as a source of truth.

File: apps/ensindexer/src/api/lib/resolution/multichain-primary-name-resolution.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    getDatasourceInAnyNamespace(config.namespace, DatasourceNames.ReverseResolverOptimism),
+    getDatasourceInAnyNamespace(config.namespace, DatasourceNames.ReverseResolverArbitrum),
+    getDatasourceInAnyNamespace(config.namespace, DatasourceNames.ReverseResolverScroll),
+  ]
+    .filter((ds) => ds !== undefined)
+    .map((ds) => ds.chain.id),
+);
+
+/**
+ * Implements batch resolution of an address' Primary Name across the provided `chainIds`. If
```
Comment (lightwalker-eth) [inline apps/ensindexer/src/api/lib/resolution/multichain-primary-name-resolution.ts:34]:
```suggestion
 * Implements multichain resolution of an address' Primary Name across the provided `chainIds`. If
```

File: packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+      name: "🔥🔥🔥🔥🔥.eth",
+    });
+  });
+
+  it("requires selection", () => {
+    expect(() => routes.records.query.parse({})).toThrow(ZodError);
+  });
+
+  it("requires normalized name", () => {
+    expect(() => routes.records.params.parse({ name: "Vitalik.eth" })).toThrow(ZodError);
```
Comment (lightwalker-eth) [inline packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts:23]: Suggest to also test with a name that cannot be normalized. Ex: "unnormalizable|name.eth"

This tests the case that normalization throws an error. This is distinct from testing a name that is normalizable, but not already normalized. Ex: "Vitalik.eth"

File: packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  });
+
+  it("requires valid address", () => {
+    expect(() => routes.primaryName.params.parse({ address: "0xabcd", chainId: "1" })).toThrow(
+      ZodError,
+    );
+  });
+
+  it("requires valid chainId", () => {
+    expect(() => routes.primaryName.params.parse({ address: zeroAddress, chainId: "-1" })).toThrow(
```
Comment (lightwalker-eth) [inline packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts:70]: Suggest to also add a test for this route when chainId is 0. Appreciate your advice on how we handle that.

File: packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+describe("routes.primaryNames", () => {
+  it("parses valid params", () => {
+    expect(routes.primaryNames.params.parse({ address: zeroAddress })).toEqual({
+      address: zeroAddress,
+    });
+  });
+
+  it("parses valid query", () => {
+    expect(routes.primaryNames.query.parse({ chainIds: "1,10,8453" })).toEqual({
```
Comment (lightwalker-eth) [inline packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts:98]: Suggest to also add a test where there's duplicate values in the chainIds array. Appreciate your advice in how we handle that.

Comment (lightwalker-eth) [inline packages/ensnode-sdk/src/ensindexer/api/zod-schemas.test.ts:98]: When we parse the chainIds that are passed, shouldn't namespace also be a param needed to determine if the provided chainIds are valid?

File: packages/ensnode-sdk/src/ensindexer/api/zod-schemas.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  .string()
+  .pipe(z.enum(["true", "false"]))
+  .transform((val) => val === "true");
+
+const stringarray = z
+  .string()
+  .transform((val) => val.split(","))
+  .pipe(z.array(z.string().min(1)).min(1));
+
+// TODO: encoded label hashes
```
Comment (lightwalker-eth) [inline packages/ensnode-sdk/src/ensindexer/api/zod-schemas.ts:32]: My first thought is for us to try to simplify this for now. For now, we can say we might return encoded labelhashes, but we can reject them as being valid inputs in our APIs.

I'm also ok to even say we just don't support encoded labelhashes even in our responses for now and follow up with this as a future feature.

Goal: Help reduce our immediate scope.

File: packages/ensnode-sdk/src/ensindexer/api/zod-schemas.ts [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+  .string()
+  .refine(
+    (val) => {
+      try {
+        return getAddress(val) === val;
+      } catch {
+        return false;
+      }
+    },
+    { error: "Must be a valid checksummed EVM Address" },
```
Comment (lightwalker-eth) [inline packages/ensnode-sdk/src/ensindexer/api/zod-schemas.ts:54]: Did you want to require checksumming? I haven't reviewed the context yet for how we use this, but the address checksumming has generally seemed like an unfortunate complexity. Perhaps I just don't appreciate it properly?

Goal: if people pass an address into our APIs, we should handle it in a case insensitive way.

File: apps/ensindexer/src/api/lib/resolution/reverse-resolution.ts [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+const tracer = trace.getTracer("reverse-resolution");
+
+/**
+ * Implements ENS Reverse Resolution, including support for ENSIP-19 L2 Primary Names.
+ *
+ * @see https://docs.ens.domains/ensip/19/#algorithm
+ *
+ * @param address the adddress whose Primary Name to resolve
+ * @param chainId the chainId within which to resolve the address' Primary Name
```
Comment (lightwalker-eth) [inline apps/ensindexer/src/api/lib/resolution/reverse-resolution.ts:30]: Suggest to document any ideas here about the "default"

File: packages/datasources/src/namespaces.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * NOTE: this is useful when the consumer wants to retrieve a Datasource from an arbitrary namespace.
+ * If using {@link getDatasource}, with a `namespaceId: ENSNamespaceId`, the typechecker will enforce
+ * that the only valid `datasourceName` is ENSRoot (the only Datasource present in all namespaces).
+ *
+ * @param namespaceId - The ENSNamespace identifier (e.g. 'mainnet', 'sepolia', 'holesky', 'ens-test-env')
+ * @param datasourceName - The name of the Datasource to retrieve
+ * @returns The Datasource object for the given name within the specified namespace, or undefined if it does not exist
+ */
+export const getDatasourceInAnyNamespace = (
```
Comment (lightwalker-eth) [inline packages/datasources/src/namespaces.ts:61]: The way I interpreted this function when reading it being called in various other files confused me.

I interpreted "getDatasourceInAnyNamespace" to mean: "guarantee to get me the requested datasource in any namespace where it is defined, it doesn't matter which one". And therefore I also assumed this function should never return undefined. It also confused me why any namespaceId param would be passed.

Now reading the code here I see it's doing something different.

Could you suggest an alternate name for this function that might reduce the chance that others would make an incorrect interpretation as I did?

File: apps/ensindexer/src/api/lib/resolution/multichain-primary-name-resolution.ts [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+import { trace } from "@opentelemetry/api";
+
+import { resolveReverse } from "@/api/lib/resolution/reverse-resolution";
+import config from "@/config";
+import { withActiveSpanAsync } from "@/lib/auto-span";
+import { DatasourceNames, getDatasource, getDatasourceInAnyNamespace } from "@ensnode/datasources";
+
+const tracer = trace.getTracer("multichain-primary-name-resolution");
+
+const ENSIP19_SUPPORTED_CHAIN_IDS: ChainId[] = uniq(
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensindexer/src/api/lib/resolution/multichain-primary-name-resolution.ts:16]: 1. not necessary in this context, as explicitly querying the default chain id is disallowed for multichain primary names resolution
2. rpc configs. first, in ensindexer the rpc config must _already_ be defined for each of the ensip-19 chains if `reverse-resolvers` plugin is enabled, that's already handled by config validation requirements. this is required to index the standalonereverseregistrar on each chain. but that's not relevant to the point i'm making.

  a. this example is not an error, it is an expected happy path. chainId values passed in the context of the resolution apis do NOT have a requirement on the associated RPC. they are a `key` used for record resolution, and they _don't_ imply an associated RPC call to the chain in question. say we have ENSIndexer configured with `subgraph,reverse-resolvers` plugins enabled. it does have the rpc config for base (because base hosts an ensip-19 standalonereverseregistrar) but that's not relevant. it does not include the `basenames` plugin so there are no Resolver records nor resolver relationships on base chain. the user then calls `resolve/primary-name/[address]/8453` to request the address' base-specific primary name. the first `resolveForward([addr].[coinType].reverse)` will be accelerated due to ENSIP-19 logic — we can just look up the base-specific primary name for the address using the `ext_primaryNames` entity which we're indexing because `reverse-resolvers` plugin is enabled. then the second forward resolution (`resolveForward(name, { addresses: [baseCoinType] })`) only needs access to the ens root chain's rpc. because the basenames plugin is not activated, the recursive `_resolveForward` to skip the CCIP-Read is never called — CCIP-Read is performed as normal, and actual http requests are made. at no point is the rpc for base consulted. the only context in which it is consulted is if the basenames plugin is enabled, which requires an associated rpcconfig for base. you can see how this generalizes to all of the ensip-19 chains.

  b. this is also an expected happy path — the protocol doesn't care which chain the coin type represents, testnet or otherwise — it allows you to query any evm coin type. just there won't be an ensip-19 reverse resolver on mainnet for [baseSepoliaCoinType].reverse and the lookups will fall back to the `default.reverse` resolver which can return a default name for any evm address (specifically to be used by EOAs)

> The default resolver will intercept "default.reverse" and "[coinTypeAsHex].reverse" for every chainFromCoinType(coinType) > 0 and lookup names when addressAsHex is a valid EVM address.

https://docs.ens.domains/ensip/19/#default-primary-name

so my reading of that is that this is explicitly allowed, and we should allow consumers to query any valid chain id

File: apps/ensindexer/src/api/lib/resolution/multichain-primary-name-resolution.ts (hunk 9 above)
Comment (shrugs) [reply apps/ensindexer/src/api/lib/resolution/multichain-primary-name-resolution.ts:34]: this is internal note and i think accurate

File: packages/ensnode-sdk/src/ensindexer/api/zod-schemas.ts (hunk 14 above)
Comment (shrugs) [reply packages/ensnode-sdk/src/ensindexer/api/zod-schemas.ts:54]: restricts api surface, having two valid formats for an `Address` is bad, and we should standardize on exactly one (checksummed or not checksummed) and anyone passing inconsistently formatted addresses around is begging for trouble (for example, we experienced at least two issues with the subgraph storing non-checksummed addresses).

our entire application suite, with the exception of legacy subgraph behavior, is standardized on checksummed addresses, so i keep that standard here

**Review by lightwalker-eth — COMMENTED** [review]
@shrugs Once again super job here. Reviewed the changes from the last few commits here. Appreciate your follow ups in a fresh PR now that this one is merged 👍

File: packages/ensnode-sdk/README.md [hunk 18]
```diff
@@ -53,55 +53,97 @@ Resolves records for an ENS name (Forward Resolution), via ENSNode, which implem
...
+//   "1": "gregskril.eth",
+//   "10": "gregskril.eth",
+//   "8453": "greg.base.eth", // base-specific Primary Name!
+//   "42161": "gregskril.eth",
+//   "59144": "gregskril.eth",
+//   "534352": "gregskril.eth"
+// }
+
+// Resolve an address' Primary Names on specific chain Ids
+const { names } = await client.resolvePrimaryNames("0x179A862703a4adfb29896552DF9e307980D19285", {
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/ensnode-sdk/README.md:138]: ensip-19 only makes sense in the context of evm chains, not to be confused with resolver address records which support all coin types
